### PR TITLE
Add time zone support

### PR DESCRIPTION
**Pull request by nohli**
Closes https://github.com/passsy/deep_pick/issues/47 and closes https://github.com/passsy/deep_pick/issues/51 (the example is the same string as the test case for `RFC 1036`).

### CODE REVIEW

File: lib/src/pick_datetime.dart [hunk 1]
```diff
@@ -235,3 +238,49 @@ const _months = {
   'Nov': 11,
   'Dec': 12,
 };
+
+int _normalizeYear(int year) {
```
Comment (passsy) [inline lib/src/pick_datetime.dart:242]: explain and add link to [rfc2822.txt](https://www.ietf.org/rfc/rfc2822.txt)

File: test/src/pick_datetime_test.dart [hunk 2]
```diff
@@ -23,6 +23,36 @@ void main() {
...
+        const input = 'Monday, 01-Jan-20 11:53:01 +0000';
+        final time = DateTime.utc(2020, 01, 01, 11, 53, 01);
+        expect(pick(input).asDateTimeOrThrow(), time);
+      });
+
+      test('parse DateTime with timezone +0730', () {
+        const input = 'Monday, 01-Nov-21 11:53:15 +0730';
+        final time = DateTime.utc(2021, 11, 01, 19, 23, 15);
+        expect(pick(input).asDateTimeOrThrow(), time);
+      });
```
Comment (passsy) [inline test/src/pick_datetime_test.dart:54]: Move those tests into the corresponding format groups like `ISO 8601`

**Comment by passsy** [comment]
What about `ISO 8601`. Does it work there? If yes, add tests. If no, we should create an issue

**Comment by nohli** [comment]
> 8601

It's not working, I added (commented out) tests that were failing.

File: lib/src/pick_datetime.dart [hunk 3]
```diff
@@ -235,3 +247,57 @@ const _months = {
...
+    }
+  }
+  return year;
+}
+
+/// The Duration to add to a DateTime to get the correct time in UTC
+Duration _getTimeZoneOffset(String? timeZone) {
+  if (timeZone == null) return Duration.zero;
+  if (RegExp(r'^[+-]\d{4}$').hasMatch(timeZone)) {
+    final operator = timeZone[0] == '-' ? 1 : -1;
```
Comment (passsy) [inline lib/src/pick_datetime.dart:273]: `operator` is a dart keyword word, better use anything else

File: lib/src/pick_datetime.dart [hunk 4]
```diff
@@ -151,7 +151,16 @@ extension NullableDateTimePick on Pick {
...
+    final trimmedValue = value.trim();
+    final timeZone = RegExp(r'[a-zA-Z]+$').firstMatch(trimmedValue)?.group(0);
+    if (timeZone != null) {
+      final timeZoneOffset = _getTimeZoneOffset(timeZone);
+      // Remove the timezone from the string and add Z, so that it's parsed as UTC
+      final newValue =
+          '${trimmedValue.substring(0, trimmedValue.length - timeZone.length)}Z';
+      return DateTime.tryParse(newValue)?.add(timeZoneOffset);
+    }
+    return DateTime.tryParse(trimmedValue);
```
Comment (passsy) [inline lib/src/pick_datetime.dart:163]: Add comment that `DateTime.tryParse` handles timezones in format `+0000`.

File: lib/src/pick_datetime.dart [hunk 5]
```diff
@@ -151,7 +151,16 @@ extension NullableDateTimePick on Pick {
...
     if (value is! String) return null;
-    return DateTime.tryParse(value);
+    final trimmedValue = value.trim();
+    final timeZone = RegExp(r'[a-zA-Z]+$').firstMatch(trimmedValue)?.group(0);
+    if (timeZone != null) {
+      final timeZoneOffset = _getTimeZoneOffset(timeZone);
+      // Remove the timezone from the string and add Z, so that it's parsed as UTC
+      final newValue =
+          '${trimmedValue.substring(0, trimmedValue.length - timeZone.length)}Z';
+      return DateTime.tryParse(newValue)?.add(timeZoneOffset);
```
Comment (passsy) [inline lib/src/pick_datetime.dart:161]: Maybe link to an issue that darts `DateTime` doesn't handle those timezone formats?

File: test/src/pick_datetime_test.dart [hunk 6]
```diff
@@ -182,6 +182,24 @@ void main() {
...
       });
 
+      test('parse DateTime with timezone +0230', () {
+        const input = '2023-01-09T12:31:54+0230';
+        final time = DateTime.utc(2023, 01, 09, 10, 01, 54);
+        expect(pick(input).asDateTimeOrThrow(), time);
+      });
+
+      test('parse DateTime with timezone EST', () {
+        const input = ' 2023-01-09T12:31:54EST ';
```
Comment (passsy) [inline test/src/pick_datetime_test.dart:192]: Split this in two tests: 
- one for trailing space
- one for EST

File: test/src/pick_datetime_test.dart [hunk 7]
```diff
@@ -182,6 +182,24 @@ void main() {
         });
       });
 
+      test('parse DateTime with timezone +0230', () {
+        const input = '2023-01-09T12:31:54+0230';
```
Comment (passsy) [inline test/src/pick_datetime_test.dart:186]: Does `DateTime.tryParse` handle trailing spaces?

File: lib/src/pick_datetime.dart [hunk 8]
```diff
@@ -235,3 +247,57 @@ const _months = {
...
+  if (RegExp(r'^[+-]\d{4}$').hasMatch(timeZone)) {
+    final operator = timeZone[0] == '-' ? 1 : -1;
+    final hours = timeZone.substring(1, 3);
+    final minutes = timeZone.substring(3, 5);
+    return Duration(
+      hours: int.parse(hours) * operator,
+      minutes: int.parse(minutes) * operator,
+    );
+  } else {
+    final timeZoneOffset = _timeZoneOffsets[timeZone] ?? 0;
```
Comment (passsy) [inline lib/src/pick_datetime.dart:281]: Due to the fallback `?? 0` we treat every unknown time zone abbreviation as `UTC`. I expect an exception similar to what `DateTime.parse()` would throw

```
FormatException: Invalid date format
2023-01-09T12:31:54XXX
```

**Comment by passsy** [comment]
Awesome, just final touches :)

File: lib/src/pick_datetime.dart (hunk 5 above)
Comment (nohli) [reply lib/src/pick_datetime.dart:161]: Haven't found an open issue regarding any of the missing timezones
https://github.com/dart-lang/sdk/issues?q=datetime+parse+is%3Aopen

File: test/src/pick_datetime_test.dart (hunk 7 above)
Comment (nohli) [reply test/src/pick_datetime_test.dart:186]: Yes, added a test for it

File: lib/src/pick_datetime.dart (hunk 8 above)
Comment (nohli) [reply lib/src/pick_datetime.dart:281]: That means the package refuses to parse when not recognizing the time zone string. Is this really desired?
Also, the method only gets `String? timeZone`, not the whole String. Should there be a second, named, parameter for the whole String, so the exception can show it?

**Comment by passsy** [comment]
Awesome, that looks great! Finally, we can mark this feature as complete

Thanks for your time and effort 🙌

**Pull request merged**